### ISSUE TITLE
Decide needed metadata for models and specify corresponding markdown formats

### ISSUE THREAD

**Issue by fsuits** [IBM/compliance-trestle#1005]
## Issue description / feature objectives
Currently the markdown files allow user-friendly editing, but the files only capture a subset of the json functionality.  Some things would be good to specify directly in the markdown somehow, e.g. items in the OSCAL metadata (title, version) and list of controls to be included by a profile.
## Caveats / Assumptions
The split functionality already has top level info in profile.json and catalog.json, so it is natural to have generically named profile.md and catalog.md, etc., in the top level markdown directories.
## Completion Criteria
- [ ] Decide what metadata are needed for each top level model
- [ ] Determine if simple markdown structures such as lists are suitable, or if it needs to be more complex yaml
- [ ] Change all the -generate routines to create an initial top level .md file for later editing
- [ ] Change all the -assemble routines to use this metadata and go direct from markdown to json

**Comment by fsuits** [comment]
As an example, for profile the list of imports is needed.
Each import requires:
- an href
- optional include-all flag
- include-controls -> list of controls
- exclude-controls -> list of controls

Even this is a subset of profile functionality, but it may be sufficient for most use-cases.  trestle currently does not support pattern matching for control inclusion - for example.

Is this complexity enough to use yaml instead of markdown?

**Comment by vikas-agarwal76** [comment]
For catalog, in the main markdown directory for catalog, we can have a catalog.md file containing the metadata associated with it.

There are 4 required fields in metadata, out those 2 needs to be specified in the markdown, and 2 can be autogenerated by trestle -
- title: should be in catalog.md
- last-modified: auto-populated by trestle while generating catalog.json
- version: should be in catalog.md
- oscal-version: auto-populated by trestle while generating catalog.json

The group ids are already captured currently as name of the directory containing the controls. Sub-groups can also be specified in a similar way using sub-directories. The group title is currently part of each control markdown file inside the group directory. For this to work, all controls for a group should have the same title in the control markdown, else it will result in an error. Same with sub-group titles also. Only subgroups can be supported this way, and not sub-controls. If sub-controls are required to be supported then some other mechanism need to be devised.

A template file should be used to specify the structure of control markdown for adding new controls, including control description (parts/subparts), guidance, parameters, any props, etc. Parameters and props may be better captured in yaml header and others as markdown content.

**Comment by vikas-agarwal76** [comment]
For profile, we can have a profile.md file in the top level markdown directory having profile title and version metadata (as described above for catalog). In addition to this a profile also needs to specify imports.

There can be one on more sections in profile.md corresponding to imports. Each import section should capture -
- href: to specify catalog or profile URL
- a list of control ids to be imported

We may not support include-all, exclude controls, or patterns. This should fine for the current requirements.

**Comment by fsuits** [comment]
I'm leaning toward yaml so it can handle more general structures and we don't need to specify a specific format or items supported.

Profile needs multiple imports and potentially complicated include/exclude structures.  As long as the dictionaries can be applied unambiguously to the model it should be ok.  Error otherwise.

backmatter is another potentially important piece and I think it would also work better with yaml than markdown.

A key point is that catalog assemble does not require a catalog.json file, but profile assemble does require one.

Also, ssp assemble doesn't require an existing ssp json, but it will load and modify it if it exists.  It uses the existing ssp to extract the components, and will set the import_profile href properly.  Otherwise the import_profile is REPLACE_ME.

In summary, current behavior allows loading all needed info from existing json for profile-assemble and ssp-assemble, but catalog-assemble currently does not check for an existing catalog json file.  A simple way to get around all this is to allow an optional catalog.json file with metadata filled in.

catalog.yaml ->
```yaml
metadata:
  title: "the title"
  version: "1.001"
```

profile.yaml ->
```yaml
metadata:
  title: "the title"
  version: "1.001"
imports:
  -
    href: my_url1
    include-all: true
  -
    href: my_url2
    include:
      - ac-1
      - ac-2
    exclude:
     - ac-3
     - ac-4
merge:
  combine:
      method: merge
  as-is: true
```

**Comment by fsuits** [comment]
Yet another option is to let all this happen in trestle create, and allow it an optional yaml file to inject into the created json model.  Then allow catalog assemble an optional catalog.json file.

But it isn't a huge burden just to use the normal trestle create and then edit the json file.

Either way - I think catalog assemble should allow an optional json file to use as the base for adding groups and controls.

**Comment by enikonovad** [comment]
Using YAML would be beneficial for us just because it is a well-defined format that user has to comply with. The conversion between JSON and YAML is also much easier. Whereas markdown is pretty much a free form text where nothing stops the user of doing something unexpected.

On another hand markdown is more "user-friendly" if the user does not know YAML structure. Fixing the user-related errors could be tricky and might made such "YAML" unreadable (we can of course add `git push` check to validate YAML structure). The big advantage of markdown here is that user can preview it directly in the Github to see if it is valid.

Representing a Profile imports in markdown could have an identical structure to the YAML.

Imports:
- Import 1:
    - href: some_url
    - include-all: true

- Import 2:
    - href: some_url
    - include:
        - ac-1
        - ac-2
    - exclude:
        - ac-3


We can also use md tables for such lists:
| include | exclude | 
|---|---|
| ac-1, ac-2 | ac-3, ac-4 |

**Comment by fsuits** [comment]
Decision was made per #1008 simply to provide a reference json catalog when doing catalog assemble.  This means all -assemble tools will have a reference json file.  Some editing of json will be needed in the overall editing cycle to provide metadata and other items but the edits should be relatively straightforward and rare.
